internal/resources/google/network_egress_internal.go

```diff
@@ -0,0 +1,306 @@
+				},
+			},
+		})
+	}
+
+	regionsData := r.getEgressRegionsData(prefixName, egressResourceType)
+	usageFiltersData := r.getEgressUsageFiltersData(egressResourceType)
+
+	for _, regData := range regionsData {
+		gRegion := regData.gRegion
+		usageKey := regData.usageKey
+
+		// TODO: Reformat to use tier helpers.
+		var usage float64
+		var used float64
+		var lastEndUsageAmount float64
+		usage = GetFloatFieldValueByUsageTag(usageKey, *r)
+
+		for idx, usageFilter := range usageFiltersData {
+			usageName := usageFilter.usageName
+			endUsageAmount := usageFilter.usageNumber
+			var quantity *decimal.Decimal
+			if endUsageAmount != 0 && usage >= endUsageAmount {
+				used = endUsageAmount - used
+				lastEndUsageAmount = endUsageAmount
+				quantity = decimalPtr(decimal.NewFromFloat(used))
+			} else if usage > lastEndUsageAmount {
+				used = usage - lastEndUsageAmount
+				lastEndUsageAmount = endUsageAmount
+				quantity = decimalPtr(decimal.NewFromFloat(used))
+			}
+			var usageFilter string
+			if endUsageAmount != 0 {
+				usageFilter = fmt.Sprint(endUsageAmount)
+			} else {
+				usageFilter = ""
+			}
+			if quantity == nil && idx > 0 {
+				continue
+			}
+			var apiRegion *string
+			if regData.fixedRegion != "" {
+				apiRegion = strPtr(regData.fixedRegion)
+			} else {
+				apiRegion = r.getEgressAPIRegionName(region, egressResourceType)
+			}
+			var name string
+			if usageName != "" {
+				name = fmt.Sprintf("%v (%v)", gRegion, usageName)
+			} else {
+				name = fmt.Sprintf("%v", gRegion)
+			}
+			attributeFilters := make([]*schema.AttributeFilter, 0)
+			if regData.apiDescriptionRegex != "" {
+				attributeFilters = append(attributeFilters, &schema.AttributeFilter{Key: "description", ValueRegex: strPtr(regData.apiDescriptionRegex)})
+			} else {
+				attributeFilters = append(attributeFilters, &schema.AttributeFilter{Key: "description", Value: strPtr(regData.apiDescription)})
+			}
+			resource.CostComponents = append(resource.CostComponents, &schema.CostComponent{
+				Name:            name,
+				Unit:            "GB",
+				UnitMultiplier:  decimal.NewFromInt(1),
+				MonthlyQuantity: quantity,
+				ProductFilter: &schema.ProductFilter{
+					Region:           apiRegion,
+					VendorName:       strPtr("gcp"),
+					Service:          r.getEgressAPIServiceName(egressResourceType),
+					AttributeFilters: attributeFilters,
+				},
+				PriceFilter: &schema.PriceFilter{
+					EndUsageAmount: strPtr(usageFilter),
+				},
+			})
+		}
+	}
+
+	return resource
+}
+
+func (r *NetworkEgressUsage) doesEgressIncludeSameContinent(egressResourceType EgressResourceType) bool {
+	switch egressResourceType {
+	case ComputeExternalVPNGateway, ComputeVPNGateway:
+		return false
+	default:
+		return true
+	}
+}
+
+func (r *NetworkEgressUsage) getEgressRegionsData(prefixName string, egressResourceType EgressResourceType) []*egressRegionData {
+	switch egressResourceType {
+	case ComputeVPNGateway:
```

Comment:
Now that we have the `NetworkEgressUsage` struct I'm wondering if it enables different approaches for handling the different egress resource types.

One option could be to use composition to extend the NetworkEgressUsage type, e.g:

```go
type ComputeVPNGatewayNetworkEgressUsage struct {
  NetworkEgressUsage
}

func (r *ComputeVPNGatewayNetworkEgressUsage) getEgressRegionsData...
```

Another option could be to have an interface for these and pass it to this struct instead of `egressResourceType`, e.g:

```go
type EgressResource interface {
  includesSameContinent() bool
  getEgressRegionsData []*EgressRegionData
}
```

What do you think?